src/scanner.c

```diff
@@ -143,7 +145,87 @@ bool tree_sitter_rust_external_scanner_scan(void *payload, TSLexer *lexer,
 
   if (lexer->lookahead == '/') {
     advance(lexer);
+
+    if ((valid_symbols[LINE_COMMENT] || valid_symbols[DOC_COMMENT]) && lexer->lookahead == '/') {
+      advance(lexer);
+
+      bool started_with_slash = lexer->lookahead == '/';
+      switch (lexer->lookahead) {
+        case '!':
+        case '/': {
+          advance(lexer);
+
+          // If three consecutive slashes were seen and this is the fourth one,
+          // the line turns back to a normal comment.
+          // The above rule does not apply for "//!" which is also a doc
+          // comment, hence why it is relevant to track started_with_slash.
+          if (started_with_slash == false || lexer->lookahead != '/') {
+            lexer->result_symbol = DOC_COMMENT;
+
+            while (true) {
+              while (true) {
+                switch (lexer->lookahead) {
+                  case '\n': {
+                    goto finished_doc_comment_line;
+                  }
+                  case 0: {
+                    goto doc_comment_exit;
+                  }
+                  default: {
+                    advance(lexer);
+                  }
+                }
+              }
+
+finished_doc_comment_line:
+
+              lexer->mark_end(lexer);
+              advance(lexer);
+              if (lexer->lookahead == '/') {
+                advance(lexer);
+                if (lexer->lookahead == '/') {
+                  advance(lexer);
+                  if (started_with_slash) {
+                    if (lexer->lookahead == '/') {
+                      advance(lexer);
+                      // If a fourth slash is found, the line turns back to a normal comment
+                      if (lexer->lookahead == '/') {
+                        break;
+                      }
+                    } else {
+                      break;
+                    }
+                  } else if (lexer->lookahead != '!') {
+                    break;
+                  }
+                } else {
+                  break;
+                }
+              } else {
+                break;
+              }
+            }
+          }
+
+          break;
+        }
+      }
+
+doc_comment_exit:
+
+      // Might have already processed a doc comment in the loop above
+      if (lexer->result_symbol != DOC_COMMENT) {
+        lexer->result_symbol = LINE_COMMENT;
+        while (lexer->lookahead != '\n' && lexer->lookahead != 0) {
```

Comment:
Maybe worth mentioning that Rust requires lines to end with `\n` according to https://doc.rust-lang.org/reference/tokens.html#string-literals, thus `\r` doesn't need to be checked for those loops.

Reply:
That seems to be only for string literals though? I've certainly had `.rs` files with CRLF endings before

Reply:
https://doc.rust-lang.org/reference/comments.html#doc-comments

> Isolated CRs (\r), i.e. not followed by LF (\n), are not allowed in doc comments.

So it will never be standalone, but can be a part of a line ending.

Reply:
> I've certainly had .rs files with CRLF endings before

What I am trying to imply by that explanation is that even in the case of CRLF the line will still **end** with `\r\n`, not `\r` alone, therefore checking only for `\n` should detect all line endings.